sound/soc/sof/amd/acp.c

```diff
@@ -413,7 +413,7 @@ int amd_sof_acp_probe(struct snd_sof_dev *sdev)
 
 	sdev->ipc_irq = pci->irq;
 	ret = request_threaded_irq(sdev->ipc_irq, acp_irq_handler, acp_irq_thread,
-				   IRQF_SHARED, "AudioDSP", sdev);
+				   IRQF_SHARED | IRQF_ONESHOT, "AudioDSP", sdev);
```

Comment:
@ajitkupandey I'm not sure about this. Looking at the implementation of ONESHOT, it clearly stated that it should not be used with SHARED. have a look here: https://github.com/thesofproject/linux/commit/b25c340c195447afb1860da580fe2a85a6b652c5

Reply:
But I see other instances in the kernel that use both together. Not sure if this was addressed later on. Better to confirm

Reply:
I have an alternative suggestion: the acp_irq_handler seems to read the DSP registers unconditionallly, and any value will trigger the thread.

Shouldn't you verify that the source of the interrupt is indeed the DSP?

Likewise, if the interrupt is shared and there is a different handler, it's got to avoid dealing with an interrupt if it's from the wrong source.

Reply:
i also wonder if the acp_irq_thread() is correct in its implementation. if it deals with two separate events, you would want to deal with EXTERNAL_INTR and  ACP_DSP_SW_INTR in the same call, and return IRQ_HANDLED after handling BOTH sources, not just the first. 

````diff
diff --git a/sound/soc/sof/amd/acp.c b/sound/soc/sof/amd/acp.c
index fe9b7dc5bc86..88e7a72914cf 100644
--- a/sound/soc/sof/amd/acp.c
+++ b/sound/soc/sof/amd/acp.c
@@ -274,12 +274,13 @@ static irqreturn_t acp_irq_thread(int irq, void *context)
 {
        struct snd_sof_dev *sdev = context;
        unsigned int val;
+       irqreturn_t irq_status = IRQ_NONE;
 
        val = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_EXTERNAL_INTR_STAT);
        if (val & ACP_SHA_STAT) {
                /* Clear SHA interrupt raised by PSP */
                snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_EXTERNAL_INTR_STAT, val);
-               return IRQ_HANDLED;
+               irq_status = IRQ_HANDLED;
        }
 
        val = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_DSP_SW_INTR_STAT);
@@ -287,10 +288,10 @@ static irqreturn_t acp_irq_thread(int irq, void *context)
                sof_ops(sdev)->irq_thread(irq, sdev);
                val |= ACP_DSP_TO_HOST_IRQ;
                snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_DSP_SW_INTR_STAT, val);
-               return IRQ_HANDLED;
+               irq_status = IRQ_HANDLED;
        }
 
-       return IRQ_NONE;
+       return irq_status;
 };
 
 static irqreturn_t acp_irq_handler(int irq, void *dev_id)
````

Reply:
Thanks for suggestions , I agree we need to tweak our irq handler to handle incoming irq's in better way. For now I'll update this PR with only one commit (semaphore one) and take these irq handling review comments separately via different PR.